File: ruby/observer/WeatherRama/observer/displays/forecast_display.rb
require './observer/display'
require './observer/observer'
require './subject/weather_station'

class ForecastDisplay
  include Observer
  include Display

  def initialize(weather_station)
    @current_pressure = 29.92
    @last_pressure = 0
    @weather_station = WeatherStation.new
    @weather_station = weather_station
    @weather_station.register_observer(self)
  end

  def update(temperature, humidity, pressure)
    super
    @last_pressure = @current_pressure
    @current_pressure = pressure
    display
  end

  def display
    super
    puts "\nForecastDisplay:"
    if @current_pressure > @last_pressure
      puts "Improving weather on the way!"
    elsif @current_pressure == @last_pressure
      puts "More of the same"
    elsif @current_pressure < @last_pressure
      puts "Watch out for cooler, rainy weather"
    end
  end

end

File: cpp/observer/WeatherRama/observer/displays/ForecastDisplay.cpp
#include <iostream>

using namespace std;

class ForecastDisplay : public Observer, public Display {

public:
    explicit ForecastDisplay(WeatherStation* weatherStation) {
        this->weatherStation = weatherStation;
        this->weatherStation->registerObserver(this);
    }

    void update(float temperature, float humidity, float pressure) override {
        lastPressure = currentPressure;
        currentPressure = pressure;
        display();
    }

    void display() override {
        cout << "\nForecastDisplay:" << endl;
        if (currentPressure > lastPressure) { cout << "Improving weather on the way!\n"; }
        else if (currentPressure == lastPressure) { cout << "More of the same\n"; }
        else if (currentPressure < lastPressure) { cout << "Watch out for cooler, rainy weather\n"; }
    }

private:
    WeatherStation* weatherStation = new WeatherStation();

    float currentPressure = 29.92;
    float lastPressure{};
};

